ghcide/src/Development/IDE/GHC/ExactPrint.hs

```diff
@@ -178,30 +179,71 @@ transformM dflags ccs uri f a = runExceptT $
         let res = printA a'
         pure $ diffText ccs (uri, T.pack src) (T.pack res) IncludeDeletions
 
+
+-- | Returns whether or not this node requires its immediate children to have
+-- be parenthesized and have a leading space.
+--
+-- A more natural type for this function would be to return @(Bool, Bool)@, but
+-- we use 'All' instead for its monoid instance.
+needsParensSpace ::
+    HsExpr GhcPs ->
+    -- | (Needs parens, needs space)
+    (All, All)
+needsParensSpace HsLam{}         = (All False, All False)
+needsParensSpace HsLamCase{}     = (All False, All False)
+needsParensSpace HsApp{}         = mempty
+needsParensSpace HsAppType{}     = mempty
+needsParensSpace OpApp{}         = mempty
+needsParensSpace HsPar{}         = (All False, All False)
+needsParensSpace SectionL{}      = (All False, All False)
+needsParensSpace SectionR{}      = (All False, All False)
+needsParensSpace ExplicitTuple{} = (All False, All False)
+needsParensSpace ExplicitSum{}   = (All False, All False)
+needsParensSpace HsCase{}        = (All False, All False)
+needsParensSpace HsIf{}          = (All False, All False)
+needsParensSpace HsMultiIf{}     = (All False, All False)
+needsParensSpace HsLet{}         = (All False, All False)
+needsParensSpace HsDo{}          = (All False, All False)
+needsParensSpace ExplicitList{}  = (All False, All False)
+needsParensSpace RecordCon{}     = (All False, All False)
+needsParensSpace RecordUpd{}     = mempty
+needsParensSpace _               = mempty
+
+
 ------------------------------------------------------------------------------
 
 {- | Construct a 'Graft', replacing the node at the given 'SrcSpan' with the
- given 'LHSExpr'. The node at that position must already be a 'LHsExpr', or
- this is a no-op.
+ given @Located ast@. The node at that position must already be a @Located
+ ast@, or this is a no-op.
+
+ You want to use 'graftExpr' instead of this function when @ast ~ 'HsExpr'@.
 -}
 graft ::
     forall ast a.
     (Data a, ASTElement ast) =>
     SrcSpan ->
     Located ast ->
     Graft (Either String) a
-graft dst = graftWithoutParentheses dst . maybeParensAST
+graft dst = graft' True dst . maybeParensAST
```

Comment:
Maybe graft should be a member of the typeclass, and then we can automatically dispatch to graftExpr. Does anyone have a strong opinion on this?

Reply:
Yeah, that's clearly the right move here.